### PR TITLE
[WIP] Implement dumping frames to videos

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Adds the feature to dump frames to videos, just like that in Dolphin. This dumps the frames in ~~native resolution (400x240 and 320x240)~~ current internal resolution of the renderer and to two separate video files, one for each screen. Adds a dependency on FFmpeg (and libavcodec needs to be version 57.48.101 or higher).
Added an action to Movie menu in citra-qt frontend, and a command line argument in citra (sdl2) frontend. 

A lot of these code was taken from Dolphin, including [AVIDump.cpp](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/AVIDump.cpp) and FindFFmpeg.cmake (Modified to detect versions of the libraries). ~~Another some of the code was taken from [danzel:screenshot](https://github.com/danzel/citra/tree/screenshot), and modified to suit the context.~~ Moved to getting pixels directly from opengl

~~To simplify use of FFmpeg in MSVC, I also added an option to use bundled binaries for FFmpeg for msvc. The PR to citra-emu/ext-windows-bin can be found at citra-emu/ext-windows-bin#18.~~ This is missing dlls, would fix.

Depends on citra-emu/build-environments#14. 

Screenshots and example videos will be added later.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4295)

### CODE REVIEW

File: src/video_core/renderer_base.h [hunk 1]
```diff
@@ -52,11 +55,25 @@ class RendererBase : NonCopyable {
...
+    bool IsDumpingFrames() {
+        return dump_frames;
+    }
+
 protected:
     EmuWindow& render_window; ///< Reference to the render window handle.
     std::unique_ptr<VideoCore::RasterizerInterface> rasterizer;
-    f32 m_current_fps = 0.0f; ///< Current framerate, should be set by the renderer
-    int m_current_frame = 0;  ///< Current frame, should be set by the renderer
+    std::unique_ptr<FrameDumper> frame_dumpers[2]; ///< Frame dumpers (one for each screen)
```
Comment (lioncash) [inline src/video_core/renderer_base.h]: `std::array`

File: src/video_core/renderer_base.h [hunk 2]
```diff
@@ -52,11 +55,25 @@ class RendererBase : NonCopyable {
 
     void RefreshRasterizerSetting();
 
+    bool StartFrameDumping(const std::string& path_top, const std::string& path_bottom,
+                           const std::string& format = "");
+    void StopFrameDumping();
+    bool IsDumpingFrames() {
```
Comment (lioncash) [inline src/video_core/renderer_base.h]: This can be a const member function.

File: src/video_core/renderer_base.h [hunk 3]
```diff
@@ -52,11 +55,25 @@ class RendererBase : NonCopyable {
...
+        return dump_frames;
```
Comment (lioncash) [inline src/video_core/renderer_base.h]:
```cpp
return dump_frames.load(std::memory_order_relaxed);
```

File: src/video_core/renderer_base.h [hunk 4]
```diff
@@ -52,11 +55,25 @@ class RendererBase : NonCopyable {
...
-    f32 m_current_fps = 0.0f; ///< Current framerate, should be set by the renderer
-    int m_current_frame = 0;  ///< Current frame, should be set by the renderer
+    std::unique_ptr<FrameDumper> frame_dumpers[2]; ///< Frame dumpers (one for each screen)
+    f32 m_current_fps = 0.0f;             ///< Current framerate, should be set by the renderers
+    int m_current_frame = 0;              ///< Current frame, should be set by the renderer
+    std::atomic_bool dump_frames = false; ///< Whether to dump frames
+    Common::Event
+        frame_dump_completed; ///< An event used to finish dumping frames on emulation stopping
+
+    u32 GetColorFillForFramebuffer(int framebuffer_index);
```
Comment (lioncash) [inline src/video_core/renderer_base.h]: This can be a const member function.

File: src/video_core/renderer_base.cpp [hunk 5]
```diff
@@ -28,3 +39,96 @@ void RendererBase::RefreshRasterizerSetting() {
...
+                    std::memcpy(px_dest, source.data(), 3);
+                }
+            }
+        } else {
+            const PAddr framebuffer_addr =
+                framebuffer.active_fb == 0 ? framebuffer.address_left1 : framebuffer.address_left2;
+            Memory::RasterizerFlushRegion(framebuffer_addr,
+                                          framebuffer.stride * framebuffer.height);
+            const u8* framebuffer_data = Memory::GetPhysicalPointer(framebuffer_addr);
+            int bpp = GPU::Regs::BytesPerPixel(framebuffer.color_format);
```
Comment (lioncash) [inline src/video_core/renderer_base.cpp]: This can be const

File: src/video_core/renderer_base.cpp [hunk 6]
```diff
@@ -28,3 +39,96 @@ void RendererBase::RefreshRasterizerSetting() {
...
+    frame_dump_completed.Reset();
+    // We swap width and height here as we are going to rotate the image as we copy it
+    // 3ds framebuffers are stored rotate 90 degrees
+    for (int i : {0, 1}) {
+        const auto& framebuffer = GPU::g_regs.framebuffer_config[i];
+        FrameDumper::FrameData frame_data{framebuffer.height, framebuffer.width};
+        LCD::Regs::ColorFill color_fill{GetColorFillForFramebuffer(i)};
+        u8* dest_buffer = frame_data.data.data();
+        if (color_fill.is_enabled) {
+            std::array<u8, 3> source;
```
Comment (lioncash) [inline src/video_core/renderer_base.cpp]:
```cpp
const std::array<u8, 3> source{{
    color_fill.color_b,
    color_fill.color_g,
    color_fill.color_r,
}};
```

File: src/video_core/frame_dumper.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    class FrameData {
+    public:
+        size_t width;
+        size_t height;
+        u32 stride;
+        std::vector<u8> data;
+
+        FrameData(size_t width_, size_t height_)
+            : width(width_), height(height_), stride(width_ * 3) {
+            data.resize(width * height * 3);
```
Comment (lioncash) [inline src/video_core/frame_dumper.h]: `, data(width * height * 3)` can be added at the end of the constructor initializer list.

File: src/video_core/frame_dumper.h [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    };
+
+    /// Initializes ffmpeg libraries. Does nothing if ffmpeg is already initialized.
+    static void Initialize();
+    void Shutdown();
+
+    bool StartDumping(const std::string& path, const std::string& format, int width, int height);
+    void StopDumping();
+    bool IsDumpingFrames();
```
Comment (lioncash) [inline src/video_core/frame_dumper.h]: This doesn't seem to be implemented anywhere.

File: src/video_core/frame_dumper.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    void AddFrame(FrameData& frame);
+
+private:
+    bool InitializeDumping();
+    void FreeResources();
+    /// Writes the encoded frame to the video file
+    void WritePacket(AVPacket& packet);
+
+    int width, height;
+    bool is_dumping = false;
```
Comment (lioncash) [inline src/video_core/frame_dumper.h]: Likewise, this seems unused.

File: src/video_core/frame_dumper.cpp [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        return;
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(58, 9, 100)
+    av_register_all();
+#endif
+    avformat_network_init();
+    initialized = true;
+}
+
+bool FrameDumper::InitializeDumping() {
+    FileUtil::CreateFullPath(dump_path);
```
Comment (lioncash) [inline src/video_core/frame_dumper.cpp]: This should probably be:

```cpp
if (!FileUtil::CreateFullPath(dump_path)) {
    return false;
}
```

File: src/citra_qt/main.h [hunk 11]
```diff
@@ -224,6 +226,10 @@ private slots:
     bool movie_record_on_start = false;
     QString movie_record_path;
 
+    // Dump frames
+    bool frame_dumping_on_start = false;
+    QString frame_dumping_path_top, frame_dumping_path_bottom;
```
Comment (lioncash) [inline src/citra_qt/main.h]: These should likely be on their own separate line.

File: src/core/hle/service/gsp/gsp_gpu.cpp [hunk 12]
```diff
@@ -7,6 +7,7 @@
 #include "common/microprofile.h"
 #include "common/swap.h"
 #include "core/core.h"
+#include "core/core_timing.h"
```
Comment (lioncash) [inline src/core/hle/service/gsp/gsp_gpu.cpp]: Why's this necessary?

File: src/core/perf_stats.cpp [hunk 13]
```diff
@@ -4,6 +4,7 @@
 
 #include <algorithm>
 #include <chrono>
+#include <cmath>
```
Comment (lioncash) [inline src/core/perf_stats.cpp]: ditto

File: src/video_core/renderer_base.cpp [hunk 14]
```diff
@@ -3,13 +3,24 @@
 // Refer to the license.txt file included.
 
 #include <memory>
+#include "common/color.h"
```
Comment (lioncash) [inline src/video_core/renderer_base.cpp]: `<cstring>` should also be included.

File: src/video_core/frame_dumper.h [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    std::string dump_path, dump_format;
+
+    AVFormatContext* format_context{};
+    AVCodecContext* codec_context{};
+    AVStream* stream{};
+    AVFrame *current_frame{}, *scaled_frame{};
```
Comment (lioncash) [inline src/video_core/frame_dumper.h]: Ditto, especially given they're pointers.

File: src/video_core/frame_dumper.h [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+private:
+    bool InitializeDumping();
+    void FreeResources();
+    /// Writes the encoded frame to the video file
+    void WritePacket(AVPacket& packet);
+
+    int width, height;
+    bool is_dumping = false;
+    std::string dump_path, dump_format;
```
Comment (lioncash) [inline src/video_core/frame_dumper.h]: These should go on their own line.

File: src/video_core/renderer_base.h [hunk 17]
```diff
@@ -52,11 +55,28 @@ class RendererBase : NonCopyable {
...
+    }
+
 protected:
     EmuWindow& render_window; ///< Reference to the render window handle.
     std::unique_ptr<VideoCore::RasterizerInterface> rasterizer;
-    f32 m_current_fps = 0.0f; ///< Current framerate, should be set by the renderer
-    int m_current_frame = 0;  ///< Current frame, should be set by the renderer
+    std::array<std::unique_ptr<FrameDumper>, 2>
+        frame_dumpers;                      ///< Frame dumpers (one for each screen)
+    f32 m_current_fps = 0.0f;               ///< Current framerate, should be set by the renderers
```
Comment (valentinvanelslande) [inline src/video_core/renderer_base.h]: Remove the `s` in `renderers`

File: src/video_core/renderer_base.h [hunk 18]
```diff
@@ -52,11 +55,28 @@ class RendererBase : NonCopyable {
...
+    int m_current_frame = 0;                ///< Current frame, should be set by the renderer
+    std::atomic_bool dump_frames = false;   ///< Whether to dump frames
+    std::atomic_bool start_dumping = false; ///< Signal to start dumping frames
+    std::atomic_bool stop_dumping =
```
Comment (valentinvanelslande) [inline src/video_core/renderer_base.h]:
```cpp
    std::atomic_bool stop_dumping =
        false; ///< Whether to stop dumping. If the renderer receive this signal, it should write an
               /// "end" marker frame to the frame dumping queue
```

File: src/video_core/renderer_base.h [hunk 19]
```diff
@@ -52,11 +55,28 @@ class RendererBase : NonCopyable {
...
+        false; ///< Whether to stop dumping. If the renderer receive this singal, it should write an
+               ///< "end" marker frame to the frame dumping queue
+    Common::Event
+        frame_dumping_stopped; ///< An event to mark the stopping of frame dumping. This is used to
+                               ///< prevent game being stopped before "end" frame is written
```
Comment (valentinvanelslande) [inline src/video_core/renderer_base.h]: `/// prevent application being stopped before "end" frame is written`
Citra can also run homebrew and other titles.

File: src/video_core/frame_dumper.h [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        size_t height;
+        u32 stride;
+        std::vector<u8> data;
+
+        FrameData(size_t width_ = 0, size_t height_ = 0, u8* data_ = nullptr);
+    };
+
+    /// Initializes ffmpeg libraries. Does nothing if ffmpeg is already initialized.
+    static void Initialize();
+    void Shutdown();
```
Comment (valentinvanelslande) [inline src/video_core/frame_dumper.h]: this is unused

**Comment by RocketRobz** [comment]
It's been a while since I tested this PR, but I recently discovered that the `Capture Screenshot` option is grayed out.
What happened to it?

**Comment by zhaowenlan1779** [comment]
@RocketRobz The screenshot feature is not in this PR, and it has been merged for a while. It is only usable when a game is running.

**Comment by RocketRobz** [comment]
@zhaowenlan1779 Ah. For some reason, I thought it was this PR that has the screenshot feature.
Weirdly, it wasn't working, even when a game is running. Downloaded the latest Canary (1169), and it's fixed now.

**Comment by zhaowenlan1779** [comment]
Superseded by #4602.

**Pull request closed without merging**